packages/zapcode-ai-python/src/zapcode_ai/__init__.py

```
@@ -152,48 +202,100 @@ def _execute_code(
-        tool_calls=tool_calls,
-    )
+        while state.get("suspended"):
+            fn_name = state["function_name"]
+            args = state["args"]
+
+            tool_def = tool_defs.get(fn_name)
+            if not tool_def:
+                raise ValueError(
+                    f"Guest code called unknown function '{fn_name}'. "
+                    f"Available: {', '.join(tool_names)}"
+                )
```

Ruff (0.15.5)

[warning] Abstract `raise` to an inner function

(TRY301)